### PR TITLE
Antiforgery middleware

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

File: src/Antiforgery/src/AntiforgeryMiddleware.cs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+            await _next(context);
+        }
+        else
+        { 
+            Log.AntiforgeryValidationFailed(_logger, message: null);
+            await WriteAntiforgeryInvalidResponseAsync(context, message: null);
+        }
+    }
+
+    private static Task WriteAntiforgeryInvalidResponseAsync(HttpContext context, string? message)
```
Comment (davidfowl) [inline src/Antiforgery/src/AntiforgeryMiddleware.cs:105]: Why is this JSON?

**Comment by Sahabee** [comment]
- [x] - - - - - [x] # _[]()[]()[]()@bratsche - [ ] # []()__[]()@___@😊

File: src/Antiforgery/src/AntiforgeryMiddleware.cs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+namespace Microsoft.AspNetCore.Antiforgery;
+
+internal sealed partial class AntiforgeryMiddleware
+{
+    private readonly IAntiforgery _antiforgery;
+    private readonly RequestDelegate _next;
+    private readonly ILogger<AntiforgeryMiddleware> _logger;
+
+    public AntiforgeryMiddleware(IAntiforgery antiforgery, RequestDelegate next, ILogger<AntiforgeryMiddleware> logger)
```
Comment (Kahbazi) [inline src/Antiforgery/src/AntiforgeryMiddleware.cs:17]: Is this middleware important enough to be included in security check in `EndpointMiddleware`?
https://github.com/dotnet/aspnetcore/blob/bc118a35bd662a48881975cfc293ee4c4bada01c/src/Http/Routing/src/EndpointMiddleware.cs#L38-L42

File: src/Antiforgery/src/AntiforgeryMiddleware.cs [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        var (success, message) = await tryValidateTask;
+        if (success)
+        {
+            Log.AntiforgeryValidationSucceeded(_logger);
+            await _next(context);
+        }
+        else
+        {
+            Log.AntiforgeryValidationFailed(_logger, message);
+            await context.Response.WriteAsJsonAsync(new ProblemDetails
```
Comment (BrennanConroy) [inline src/Antiforgery/src/AntiforgeryMiddleware.cs:82]: Any reason this doesn't set the status code to 400 or call `WriteAntiforgeryInvalidResponseAsync`?

**Comment by pranavkm** [comment]
@Tratcher pointed out that we should be applying the form limits before this middleware is executed. In MVC, this is currently implemented as a separate auth filter: https://github.com/dotnet/aspnetcore/blob/main/src/Mvc/Mvc.Core/src/RequestFormLimitsAttribute.cs, https://github.com/dotnet/aspnetcore/blob/main/src/Mvc/Mvc.Core/src/Filters/RequestFormLimitsFilter.cs.

Would we want to roll these two things in to a separate `RequestFormMiddleware`? It feels unwieldy to have two separate middlewares with specific ordering requirements.

/cc @DamianEdwards in case you wanted to chime in.

**Comment by DamianEdwards** [comment]
Is the intent here to just have this middleware validate antiforgery tokens/cookies/headers if present in the request? i.e. issuing tokens and sending them back to the client is still up to the framework or manual app code.

File: src/Antiforgery/src/AntiforgeryMiddleware.cs (hunk 1 above)
Comment (DamianEdwards) [reply src/Antiforgery/src/AntiforgeryMiddleware.cs:105]: Seems wrong to be forcibly returning JSON here. Why wouldn't we just let it throw or throw `BadHttpRequestException` or something else to enable the app to handle how this responds to the client?

Comment (Tratcher) [reply src/Antiforgery/src/AntiforgeryMiddleware.cs:105]: Or make it an event.

File: src/Antiforgery/src/Internal/DefaultAntiforgery.cs [hunk 4]
```diff
@@ -100,35 +98,43 @@ public async Task<bool> IsRequestValidAsync(HttpContext httpContext)
...
             HttpMethods.IsOptions(method) ||
-            HttpMethods.IsTrace(method))
+            HttpMethods.IsTrace(method)))
         {
             // Validation not needed for these request types.
-            return true;
+            return (true, null);
         }
 
         var tokens = await _tokenStore.GetRequestTokensAsync(httpContext);
```
Comment (Tratcher) [inline src/Antiforgery/src/Internal/DefaultAntiforgery.cs:121]: Something should catch the AntiforgeryValidationException and convert it to a 400.
https://github.com/dotnet/aspnetcore/blob/a450cb69b5e4549f5515cdb057a68771f56cefd7/src/Antiforgery/src/Internal/DefaultAntiforgeryTokenStore.cs#L62-L77

**Comment by pranavkm** [comment]
> Is the intent here to just have this middleware validate antiforgery tokens/cookies/headers if present in the request? i.e. issuing tokens and sending them back to the client is still up to the framework or manual app code.

Today, Antiforgery is implemented as a filter in MVC. There's an attribute that also acts as an auth filter. It reads the form and performs a validation. Razor Pages add an implicit antiforgery filter that is active for all non-idempotent (POST / PUT / DELETE etc) requests. We'd like to enable Antiforgery to become available outside of MVC. My goal was to continue relying on these attributes, but as metadata that indicated if the user intended to perform antiforgery validation for that endpoint.

In minimal endpoints, this might look like:

```C#
app.MapPost("/post", (IFormFile formFile) => { ... })
   .ValidateAntiforgery(); // Maybe we can infer this based on the presence of things that come from a HTTP form.
```

Generating the antiforgery token that goes in to the form is still tied to the app framework. MVC will generate if you're using the form tag helper / Html.BeginForm(). For a form post using minimal, a user can use the `IAntiforgery` service to produce it. This PR doesn't attempt to make that experience better.

**Pull request closed without merging**